botocore/client.py

```diff
@@ -114,10 +114,73 @@ def _register_retries(self, service_model):
         S3RegionRedirector(endpoint_bridge, client).register()
+        self._set_s3_addressing_style(
+            endpoint_url, client.meta.config.s3, client.meta.events)
+        # Enable accelerate if the configuration is set to to true or the
+        # endpoint being used matches one of the accelerate endpoints.
+        if self._is_s3_accelerate(endpoint_url, client.meta.config.s3):
+            # Also make sure that the hostname gets switched to
+            # s3-accelerate.amazonaws.com
+            client.meta.events.register_first(
+                'request-created.s3', switch_host_s3_accelerate)
+
+    def _set_s3_addressing_style(self, endpoint_url, s3_config, event_emitter):
+        if s3_config is None:
+            s3_config = {}
+
+        style = self._get_s3_addressing_style(endpoint_url, s3_config)
+        handler = self._get_s3_addressing_handler(
+            endpoint_url, s3_config, style)
+        if handler is not None:
+            event_emitter.register('before-sign.s3', handler)
+
+    def _get_s3_addressing_style(self, endpoint_url, s3_config):
+        # Use virtual host style addressing if accelerate is enabled or if
+        # the given endpoint url is an accelerate endpoint.
+        accelerate = s3_config.get('use_accelerate_endpoint', False)
+        if accelerate or self._is_s3_accelerate(endpoint_url, s3_config):
+            return 'virtual'
+
+        # If a particular style is configured, use it.
+        configured_style = s3_config.get('addressing_style')
+        if configured_style:
+            return configured_style
+
+    def _get_s3_addressing_handler(self, endpoint_url, s3_config, style):
+        # If virtual host style was configured, use it regardless of whether
+        # or not the bucket looks dns compatible.
+        if style == 'virtual':
+            return switch_to_virtual_host_style
+
+        # If path style is configured, no additional steps are needed. If
+        # endpoint_url was specified, don't default to virtual. We could
+        # potentially default provided endpoint urls to virtual hosted
+        # style, but for now it is avoided.
+        if style == 'path' or endpoint_url is not None:
+            return None
+
+        # For dual stack mode, we need to clear the default endpoint url in
+        # order to use the existing netloc if the bucket is dns compatible.
+        if s3_config.get('use_dualstack_endpoint', False):
+            return functools.partial(
+                fix_s3_host, default_endpoint_url=None)
+
+        # By default, try to use virtual style with path fallback.
+        return fix_s3_host
+
+    def _is_s3_accelerate(self, endpoint_url, s3_config):
+        if s3_config is not None and s3_config.get('use_accelerate_endpoint'):
+            return True
+
+        if endpoint_url is None:
+            return False
+
+        parts = urlsplit(endpoint_url).netloc.split('.')
+        return any(part == 's3-accelerate' for part in parts)
```

Comment:
It won't trigger false positives for containing bucket names since it's an equality check and not an in check. Supporting a specified endpoint url is deliberate because that's what we do now. I think we could do away with blowing out user provided endpoint urls though since we force addressing style separately and that was the real concern anyway.